Rework combobox.jsx to functional component

Gazook89 opened the pull request:
## Description

This is a rework of the Combobox component.  It is now a Functional component.  It makes a few small changes to structure (the options are all buttons now).  It also adds better keyboard navigation (which is pretty much all of the "new lines of code").  Beyond that, it is mostly the same at this point.

Right now, only the Language picker uses this component.  I started to rework it because I want to use it in the TagInput component (aka StringArrayEditor).  

Beyond whatever is suggested in review, some additional things I want to do:

- [x] add a "expand"/"open" button to the right side of the input as a button to open it (in truth, the whole input toggles the menu so this will just be a little visual indicator).
- [ ] a quick pass on accessibility issues (roles, hidden navigation)
- [ ] get the validators working with it again.

Then, as a subsequent PR, I want to work out a good way to get submenus to work.


## Related Issues or Discussions

- Related #3735

## QA Instructions, Screenshots, Recordings

Try out the Language picker, using your mouse and navigating it with the keyboard.  You should be able to:

- Tab into it and focus the input
- Space to open the menu with all options, Space again to close it
- Type any characters to open menu to filtered options
- Typing Space should still work to add whitespace, without opening/closing the menu, as long as the input isn't otherwise empty
- Esc to close the drop down
- Enter to select an option
- ArrowDown from input to open and navigate through menu, cycling to first option if reached the bottom
- ArrowUp to move through menu, focusing the input when you move past the first option
- when focus is on menu option, be able to use Backspace to move focus to input and delete characters.
- Tab out to next input closes the dropdown and moves focus.
- entering a value that filters to zero results displays a 'no matches' item.

Possibly some of these behaviors will change as i look at WCAG for comboboxes and I'll try to update this list.  

### Reviewer Checklist

*Reviewers, refer to this list when testing features, or suggest new items *

<details><summary>Copy this list</summary>

```
- [ ] Verify new features are functional
	- [ ] Tab into it and focus the input
	- [ ] Space to open the menu with all options, Space again to close it
	- [ ] Type any characters to open menu to filtered options
	- [ ] Typing Space should still work to add whitespace, without opening/closing the menu, as long as the input isn't otherwise empty
	- [ ] Esc to close the drop down
	- [ ] Enter to select an option
	- [ ] ArrowDown from input to open and navigate through menu, cycling to first option if reached the bottom
	- [ ] ArrowUp to move through menu, focusing the input when you move past the first option
	- [ ] when focus is on menu option, be able to use Backspace to move focus to input and delete characters.
	- [ ] Tab out to next input closes the dropdown and moves focus.
	- [ ] entering a value that filters to zero results displays a 'no matches' item.
- [ ] Verify old features have not broken
  - [ ] Still correctly sets the `lang` attribute and is preserved with the brew metadata
- [ ] Test for edge cases / try to break things
- [ ] Identify opportunities for simplification and refactoring
- [ ] Check for code legibility and appropriate comments
```

</details>

Gazook89 committed:
Rework combobox.jsx to functional component

This is barebones features of the combobox, but written as a functional component.  It features a text input and a div containing a list of buttons with the different options included.  It functionally updates the brew metadata props and loads them when the component initializes.

Many features of the original combobox are still missing.

Light re-styling.

Small HTML changes to metadataEditor.  Change the list of options to buttons rather than divs with onClick attributes.

Change how options are created and passed into Combobox component.

Gazook89 committed:
add handler for click outside of component

Closes the dropdown menu when clicked outside the component.

Gazook89 committed:
open dropdown if closed, and arrowed down

Gazook89 committed:
Merge branch 'master' into Functional-Combobox

Gazook89 committed:
small linting and simplifying useEffects

Gazook89 committed:
set default filteredOptions state to include all children

This allows the dropdown to contain all available children when it is first opened when otherwise it would be empty.

Gazook89 committed:
add ref to text input node.

Gazook89 committed:
small whitespace linting

small lintining

Gazook89 committed:
add keyboard navigation on keyDown

Gazook89 committed:
set a "no matches" item in menu if no matches

Creates a simple span that indicates no matches for a given input value.

Gazook89 committed:
fix a bad earlier lint.

Gazook89 committed:
Add placeholder prop, start managing focus

Gazook89 committed:
set styling of 'no matches' item

Gazook89 committed:
lint, prevent input clear on click

5e-Cleric commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,166 @@
+		setInputValue(newValue);
+		setCurrentOption(-1);
+
+		const filtered = React.Children.toArray(props.children).filter((option)=>{
+			return autoSuggest.filterOn.some((filterAttr)=>{
+				return option.props[filterAttr]?.toLowerCase().startsWith(newValue.toLowerCase());
+			});
```
I believe we can make this using `some`, making it more concise without losing readability, make the change if you like

```suggestion
const filtered = props.children.filter(({ props }) =>
  autoSuggest.filterOn.some(filterAttr =>
    props[filterAttr]?.toLowerCase().startsWith(newValue.toLowerCase())
  )
);

```

Gazook89 replied:
This is nearly the exact same as what I have unless I’m missing something or don’t know how to read a diff. 

The React.Children.toArray is important before there may be only a single child, not an array, so this would convert it to an array. Or, it could have multiple children but just not passed as an array. 

But regarding `some`, pretty sure I’m already using it the exact same way?

5e-Cleric replied:
basically, alright, then this is resolved. Did not think of the single element possibility

5e-Cleric commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,166 @@
+				}
+			}
+
+			// ArrowUp moves to the previous option
+			else if(evt.key === 'ArrowUp') {
+				evt.preventDefault();
+				const prevIndex = currentOption - 1;
+				if(prevIndex >= 0) {
+					setCurrentOption(prevIndex);
+				} else {
+					setCurrentOption(-1);
+					inputRef.current.focus();
+				}
+			}
+
+			// Escape key closes the dropdown
+			else if(evt.key === 'Escape'){
+				setCurrentOption(-1);
+				inputRef.current.focus();
+				setShowDropdown(false);
+			}
+
+			// Backspace key while menu is open still deletes characters in input
+			else if((evt.key === 'Backspace') && showDropdown){
+				inputRef.current.focus();
+			}
+
+			else if((evt.key === 'Tab')){
+				setCurrentOption(-1);
+				setShowDropdown(false);
+			}
+
+			// Prevent modifier keys from triggering dropdown (example, shift+tab or tab to move focus)
+			else if(!modifiers.includes(evt.key)) {
+				setShowDropdown(true);
+			}
+		}
+
+
+	};
```
With that many else ifs, we might benefit from using the hated switch:

```suggestion
    const handleKeyDown = (evt) => {
    const modifiers = ['Meta', 'Shift', 'Alt', 'Control', 'Tab'];
    const isArrowDown = evt.key === 'ArrowDown';
    const isArrowUp = evt.key === 'ArrowUp';
    const optionsLength = filteredOptions.length;

    if (inputFocused || currentOption >= 0) {
        switch (evt.key) {
            case ' ':
                if (inputValue === '') {
                    evt.preventDefault();
                    setShowDropdown(prev => !prev);
                }
                break;
                
                // ArrowDown moves to the next option
            case 'ArrowDown':
                evt.preventDefault();
                setShowDropdown(true);
                setCurrentOption((currentOption + 1) % optionsLength);
                break;
                
                // ArrowUp moves to the previous option
            case 'ArrowUp':
                evt.preventDefault();
                const newIndex = currentOption - 1;
                setCurrentOption(newIndex >= 0 ? newIndex : -1);
                if (newIndex === -1) inputRef.current.focus();
                break;
                
                // Escape key closes the dropdown
            case 'Escape':
                setShowDropdown(false);
                setCurrentOption(-1);
                inputRef.current.focus();
                break;
                
                // Backspace key while menu is open still deletes characters in input
            case 'Backspace':
                if (showDropdown) inputRef.current.focus();
                break;
            case 'Tab':
                setShowDropdown(false);
                setCurrentOption(-1);
                break;
                
                // Prevent modifier keys from triggering dropdown (example, shift+tab or tab to move focus)
            default:
                if (!modifiers.includes(evt.key)) setShowDropdown(true);
                break;
        }
    }
};

```

Please double check it in case i made many errors.

Gazook89 replied:
I considered it, but I opted for else if before I personally think it is more readable and understandable (don’t have to worry about “making many errors”. It doesn’t save many (any?) lives, either (though I can see I have some extra white space). 

I’ll reconsider if Switch/Case has a huge amount of support among reviewers

Gazook89 replied:
I'm going to mark this as 'resolved' for right now mostly because it's enormous amount of text in the middle of the thread.  If anyone else comments that they want Switch/Case statements rather than else/if, I'll *switch* it.

calculuschild replied:
I'll add another vote for `switch`. This seems like the perfect use case for that.

5e-Cleric commented on client/components/combobox.less:
```diff
@@ -23,16 +27,26 @@
             padding: 5px;
             cursor: default;
             margin: 0 3px;
+			box-sizing: border-box;
```
I believe our css applies this to all elements, consider changing `all:unset` if the approach is taking more lines than it should (you probably already have, but must suggest it)

Gazook89 replied:
Largely the `all: unset` is to unset everything we have currently setup for *every* `button` on the site:

```css
button {
	-webkit-transition: background-color .25s ease;
	-moz-transition: background-color .25s ease;
	-o-transition: background-color .25s ease;
	transition: background-color .25s ease;
	display: inline-block;
	padding: .6em 1.2em;
	cursor: pointer;
	background-color: #27AE60;
	font-family: 'Open Sans',sans-serif;
	font-size: .8em;
	font-weight: 800;
	color: white;
	text-decoration: none;
	text-transform: uppercase;
	border: none;
	outline: none;
}
```

Probably it'd be ideal to change this styling to be based on *some* class, rather than styling every button with a very heavy-handed style.  As we start to use actual button elements rather than just `div` elements with `onClick` attributes, we are going to want to style those buttons in certain contexts ("navbar", "snippetbar", "viewbar", etc).

If I made that change in this PR, though, the PR would quickly spread to cover changes in many more files.  So probably left as it's own thing for now.  Instead, I'll just do `all: unset` and then build the style from the ground up.

Without putting much thought on it, I think the current "default button" styling should apply to the class name `call-to-action`, though I bet there are better names.

5e-Cleric commented on client/components/combobox.less:
```diff
@@ -23,16 +27,26 @@
+			margin: 0;
             //border-bottom: 1px solid darkgray;
             &:hover {
+                background-color: rgb(223, 221, 221);
+			}
+			&:focus {
                 filter: brightness(120%);
                 background-color: rgb(163, 163, 163);
             }
```
Overall, i'd say keep the colors in hex whenever possible, less characters, same legibility IMO

Gazook89 replied:
Agreed, thanks for pointing that out.  I think it's a remnant from when i was trying to do tricky color manipulations here.  I went further and lightened the grays a bit more too.  This will be resolved when I push up my changes, so marking it now.

5e-Cleric commented on client/homebrew/editor/metadataEditor/metadataEditor.jsx:
```diff
@@ -240,16 +240,15 @@ const MetadataEditor = createClass({
+		const langOptions = _.map(langCodes.sort(), (code, index)=>{
+			const localName = new Intl.DisplayNames([code], { type: 'language' });
+			const englishName = new Intl.DisplayNames('en', { type: 'language' });
+			return <button className='item' title={`${englishName.of(code)}`} key={`${index}`} data-value={`${code}`} data-detail={`${localName.of(code)}`}>
```
can't all of these be written as `title={englishName.of(code)}`?

Gazook89 replied:
yes, good point, thanks -- fixed

5e-Cleric reviewed (commented):
Overall looks fine, i made more suggestions than i thought, sorry for making them separate.

Gazook89 committed:
add dropdown button and styling

Gazook89 committed:
Open dd menu with Space when text is selected

This makes it easier to open the menu with Space when you tab into the input, since the text gets highlighted (and would otherwise get deleted/replaced by ' ').

Gazook89 committed:
change rgb to hex, some colors

lightened up the grays on focus and hover to leave the writing slightly more legible.

Gazook89 committed:
clean up some strings in metadataEditor

Gazook89 committed:
Add aria attributes to combobox.

calculuschild commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,167 @@
+	const optionRefs = useRef([]);
+	const componentRef = useRef(null);
+
+	useEffect(()=>{
```
Here, you are re-defining `handleClickOutside()` every time `showDropdown` changes value. I don't think this is necessary; you can define the function outside of `useEffect`. Same for the other functions below.

Gazook89 replied:
Got it, good point on this one.  done.

calculuschild commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,167 @@
-		if(this.refs.dropdown && !this.refs.dropdown.contains(e.target)) {
-			this.handleDropdown(false);
+
+		document.addEventListener('pointerdown', handleClickOutside);
```
This is also adding a new event listener every time `showDropdown` changes.

If you want to add a listener once when the component first mounts, change the `[showDropdown]` to an empty array `[]`

Gazook89 replied:
If just change this to onMount `[]`, it doesn't work.  But it works if i remove the `showDropdown` condition in the handler:

```
const handleClickOutside = (evt)=>{
		console.log(componentRef.current, showDropdown)
		if(componentRef.current && !componentRef.current.contains(evt.target)) {
			setShowDropdown(false);
		}
	};
```

So that's what i'll do.

calculuschild replied:
Without any dependency array it now adds a new listener on *every* render, which is even worse.

Let's see if we can figure out why `[]` didn't work

calculuschild commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,167 @@
+		if(inputFocused || (currentOption >= 0)) {
+			const optionsLength = filteredOptions.length;
+
+			if(evt.key === ' ' && (inputValue === '' || (inputRef.current.selectionStart === 0 && inputRef.current.selectionEnd === inputValue.length))){
+				evt.preventDefault();
+				setShowDropdown(!showDropdown);
+			}
```
What is the purpose of this one?

calculuschild replied:
Sorry, missed it in your original comment. Ignore this.

calculuschild commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,167 @@
+		return ()=>{document.removeEventListener('pointerdown', handleClickOutside);};
+	}, [showDropdown]);
+
+	useEffect(()=>{
+		onSelect(inputValue);
+		// handleInputChange({ target: { value: inputValue } });
+	}, [inputValue]);
```
`useEffect` is not usually not needed for this type of case. Here, we trigger two renders, first when `inputValue` changes, and then again when `onSelect()` changes something in the parent `metaDataEditor`.

Instead of setting `inputValue` and then handling side effects later in `useEffect`, both `setInputValue` and calling `onSelect()` should happen together in a handler function, probably in your `handleInputChange()` function.

calculuschild commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,167 @@
+	// Handle keyboard navigation
+	const handleKeyDown = (evt)=>{
+		const modifiers = ['Meta', 'Shift', 'Alt', 'Control', 'Tab'];
+		if(inputFocused || (currentOption >= 0)) {
```
If we are already focused, why do we need to keep setting focus for the different keys below?

calculuschild commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,167 @@
+		// handleInputChange({ target: { value: inputValue } });
+	}, [inputValue]);
+
+	useEffect(()=>{
```
Similar comment here. Changing `currentOption` *and* the side effect of focusing the element should happen inside of a handler function, not inside `useEffect`.

See https://react.dev/learn/you-might-not-need-an-effect

Gazook89 committed:
update props init to keep props object

no destructuring.

Gazook89 commented on client/components/combobox.jsx:
```diff
@@ -1,128 +1,167 @@
-			value        : '',
-			options      : [...this.props.options],
-			inputFocused : false
+const Combobox = ({ onSelect, onEntry, autoSuggest = { filterOn: ['data-value'], suggestMethod: 'includes', clearAutoSuggestOnClick: false }, ...props })=>{
```
donezo.

Gazook89 committed:
move handler function to top level, out of effect

And removes the dependence on showDropdown of that effect.

Gazook89 commented:
Thanks for reviewing.  I'm going to leave some comments unaddressed because I'm in the middle of refiguring out how focus is managed and a few small behavior changes as part of the "align with WCAG" portion of this.  I didn't think I'd get two code reviews so quick :)  but they were still valuable and I changed a few things as suggested.

Gazook89 commented:
Going to close this and come back to it, maybe, with smaller & better commits and goals.
